### PR TITLE
Add more grad functions

### PR DESCRIPTION
**Pull request by hjchen2**
这个PR依赖了 #4673 。

### CODE REVIEW

File: oneflow/core/autograd/gradient_funcs/broadcast_binary_ops.cpp [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+namespace one {
+
+namespace {
+
+class ReduceSumLikeModule {
+ public:
+  ReduceSumLikeModule(const std::string& op_name) {
+    identity_op_ = op_expr_helper::IdentityOp(op_name).GetPtrOrThrow();
+    reshape_like_op_ = op_expr_helper::ReshapeLikeOp(op_name).GetPtrOrThrow();
+    reduce_sum_like_op_ = op_expr_helper::ReduceSumLikeOp({-1}, op_name).GetPtrOrThrow();
```
Comment (strint) [inline oneflow/core/autograd/gradient_funcs/broadcast_binary_ops.cpp:32]: 三个op同名是不是不太好？

File: oneflow/core/autograd/gradient_funcs/broadcast_binary_ops.cpp [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+ private:
+  std::shared_ptr<OpExpr> identity_op_;
+  std::shared_ptr<OpExpr> reshape_like_op_;
+  std::shared_ptr<OpExpr> reduce_sum_like_op_;
+};
+
+}  // namespace
+
+class BroadcastBinary : public OpExprGradFunction<OpExprInterpState> {
```
Comment (strint) [inline oneflow/core/autograd/gradient_funcs/broadcast_binary_ops.cpp:66]: BroadcastBinary -> BroadcastBinaryGrad 更准确点？

**Review by strint — COMMENTED** [review]
本pr新增是重定义的grad func实现，通过后面的CI op测试来保证正确性就好。

File: oneflow/core/autograd/gradient_funcs/broadcast_binary_ops.cpp [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+ public:
+  BroadcastBinary() = default;
+  virtual ~BroadcastBinary() = default;
+
+  virtual Maybe<void> Init(const OpExpr& op) {
+    const auto* fw_op_expr = dynamic_cast<const UserOpExpr*>(&op);
+    CHECK_NOTNULL_OR_RETURN(fw_op_expr);
+    op_name_ = fw_op_expr->op_name();
```
Comment (wyg1997) [inline oneflow/core/autograd/gradient_funcs/broadcast_binary_ops.cpp:74]:
```suggestion
    op_name_ = GradientOpName(fw_op_expr->op_name());
```

实现的grad function的op_name好像没有一个统一的规则

Comment (wyg1997) [reply oneflow/core/autograd/gradient_funcs/broadcast_binary_ops.cpp:74]: 后续有时间可以整理一下op_name这块，最终可以把名字带进FunctionNode中

Comment (hjchen2) [reply oneflow/core/autograd/gradient_funcs/broadcast_binary_ops.cpp:74]: 其实有一个规则的，通常就是在当前op name基础上加一个".grad"后缀，如果对应多个op，就把op name基础上再加一个当前op的一个描述。

Comment (wyg1997) [reply oneflow/core/autograd/gradient_funcs/broadcast_binary_ops.cpp:74]: 是，只是我看有的grad function加了，有的没有。如果统一有名字的话，op_name应该是父类的成员吧？

Comment (hjchen2) [reply oneflow/core/autograd/gradient_funcs/broadcast_binary_ops.cpp:74]: 都需要加的，可能有些地方遗漏了。这个op_name_是缓存的一个前向op name，不是后向的，后向name需要GradientOpName生成

Comment (wyg1997) [reply oneflow/core/autograd/gradient_funcs/broadcast_binary_ops.cpp:74]: 哦哦明白了

File: oneflow/core/autograd/gradient_funcs/broadcast_binary_ops.cpp (hunk 1 above)
Comment (hjchen2) [reply oneflow/core/autograd/gradient_funcs/broadcast_binary_ops.cpp:32]: 其实这是OK的，因为这三个op正常情况下只会使用其中一个，但确实可能存在dynamic的情况，导致每次调用的op不一样。这里我改一下。

File: oneflow/core/autograd/gradient_funcs/broadcast_binary_ops.cpp (hunk 2 above)
Comment (hjchen2) [reply oneflow/core/autograd/gradient_funcs/broadcast_binary_ops.cpp]: done

File: oneflow/core/autograd/gradient_funcs/broadcast_binary_ops.cpp (hunk 1 above)
Comment (hjchen2) [reply oneflow/core/autograd/gradient_funcs/broadcast_binary_ops.cpp]: done

**Review by strint — APPROVED** [review]

File: oneflow/core/autograd/gradient_funcs/bias_add.cpp [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    return Maybe<void>::Ok();
+  }
+
+  Maybe<void> Apply(const BiasAddInterpState* ctx, const TensorTuple& out_grads,
+                    TensorTuple* in_grads) const override {
+    const int64_t num_axes = out_grads.at(0)->shape()->NumAxes();
+    in_grads->resize(2);
+    if (ctx->bias_requires_grad) {
+      std::vector<int32_t> reduce_axes_vec;
+      for (int i = 0; i < num_axes; ++i) {
```
Comment (lixinqi) [inline oneflow/core/autograd/gradient_funcs/bias_add.cpp:59]:
```c++
reduce_axes_vec.reserve(num_axes);
```
避免std::vector的隐式resize

File: oneflow/core/autograd/gradient_funcs/cast.cpp [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+#include "oneflow/core/framework/dtype.h"
+#include "oneflow/core/framework/op_builder.h"
+#include "oneflow/core/framework/op_interpreter/op_interpreter_util.h"
+#include "oneflow/core/framework/op_expr.h"
+#include "oneflow/core/framework/op_expr_helper.h"
+
+namespace oneflow {
+namespace one {
+
+class Cast : public OpExprGradFunction<OpExprInterpState> {
```
Comment (lixinqi) [inline oneflow/core/autograd/gradient_funcs/cast.cpp:27]: OpExprGradFunction的设计很用心。

File: oneflow/core/autograd/gradient_funcs/bias_add.cpp (hunk 4 above)
Comment (hjchen2) [reply oneflow/core/autograd/gradient_funcs/bias_add.cpp:59]: 在新的PR #4699 中修改

File: oneflow/core/autograd/gradient_funcs/normalization.cpp [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    // The normalization op has 5 inputs which are x, moving_mean, moving_variance, gamma and beta.
+    in_grads->resize(5);
+    in_grads->at(3) = results->at(1);  // gamma_diff;
+    in_grads->at(4) = results->at(2);  // beta_diff
+    if (ctx->is_training) {
+      in_grads->at(0) = results->at(0);
+      return Maybe<void>::Ok();
+    }
+
+    DimVector dim_vec;
```
Comment (wyg1997) [inline oneflow/core/autograd/gradient_funcs/normalization.cpp:113]:
```C++
DimVector dim_vec(x->shape()->NumAxes());
```

File: oneflow/core/autograd/gradient_funcs/split_like.cpp [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+Maybe<void> SplitLike::Apply(const SplitLikeInterpState* ctx, const TensorTuple& out_grads,
+                             TensorTuple* in_grads) const {
+  in_grads->resize(1);
+  if (!ctx->requires_grad) { return Maybe<void>::Ok(); }
+
+  CHECK_EQ_OR_RETURN(out_grads.size(), zero_like_ops_.size());
+  const auto& saved_tensors = ctx->SavedTensors();
+  TensorTuple inputs;
```
Comment (wyg1997) [inline oneflow/core/autograd/gradient_funcs/split_like.cpp:83]:
```
TensorTuple inputs(out_grads.size());
```
